### ISSUE TITLE
replaceById() error message

### ISSUE THREAD

**Issue by PhilippeCorreges** [mean-expert-official/loopback-sdk-builder#283]
#### What type of issue are you creating?
- [ ] Bug
- [ ] Enhancement
- [X] Question

#### What version of this module are you using?
- [ ] 2.0.10 (Stable)
- [X] 2.1.0-rc.n (2.1 Release Candidate n)
- [ ] Other

Write other if any:

#### Please add a description for your issue:
Hello,
When i try to do a replaceById() I received the following error message:
`Object
message
:
"Cannot call Link.replaceById(). The replaceById method has not been setup. The PersistedModel has not been correctly attached to a DataSource!"
name
:
"Error"
stack
:
"Error: Cannot call Link.replaceById(). The replaceById method has not been setup. The PersistedModel has not been correctly attached to a DataSource!↵    at throwNotAttached (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback/lib/persisted-model.js:68:11)↵    at Function.replaceById (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback/lib/persisted-model.js:558:5)↵    at SharedMethod.invoke (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/shared-method.js:263:25)↵    at HttpContext.invoke (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/http-context.js:389:12)↵    at phaseInvoke (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/remote-objects.js:654:9)↵    at runHandler (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/lib/phase.js:135:5)↵    at iterate (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/node_modules/async/lib/async.js:146:13)↵    at Object.async.eachSeries (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/node_modules/async/lib/async.js:162:9)↵    at runHandlers (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/lib/phase.js:144:13)↵    at iterate (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/node_modules/async/lib/async.js:146:13)↵    at /Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/node_modules/async/lib/async.js:157:25↵    at /Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/node_modules/async/lib/async.js:154:25↵    at execStack (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/remote-objects.js:493:7)↵    at RemoteObjects.execHooks (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/remote-objects.js:497:10)↵    at phaseBeforeInvoke (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/remote-objects.js:650:10)↵    at runHandler (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/lib/phase.js:135:5)"
status
:
500`

Is this something known ?

My code for that is:

` submit(): void{
        this.link.replaceById(this.links.linkId).subscribe((linkInstance: Link) =>{});
    };`
with:

` constructor(private link: LinkApi,
                private route: ActivatedRoute,
                private router: Router,
                )
    {}` 

in the component.

Is there something wrong with what I wrote ? The error message sounds strange to me.

Thanks for any input.

**Comment by jonathan-casarrubias** [comment]
Hi @PhilippeCorreges thanks for reaching out.

It seems that your model does not have a datasource for that model, at least that is what the error message says, it may be missing the configuration within model-config.json file

**Comment by PhilippeCorreges** [comment]
Hi @jonathan-casarrubias 
Thanks for the answer but unfortunately the model has a datasource ...
I will come back once I have more details.

Regards,

**Comment by jonathan-casarrubias** [comment]
Sure, maybe a malformed request url could cause this, but this seems to be an error thrown by loopback itself

**Comment by PhilippeCorreges** [comment]
ho @jonathan-casarrubias 
Maybe it is coming from the fact that I have a relationship between Category Model and Links of type : 
`  "relations": {
    "links": {
      "type": "hasMany",
      "model": "Link",
      "foreignKey": "categoryId"
    }`

Link is like : 
`{
  "name": "Link",
  "plural": "Links",
  "base": "PersistedModel",
  "idInjection": true,
  "options": {
    "validateUpsert": true
  },
  "properties": {
    "linkName": {
      "type": "string",
      "required": true
    },
    "linkURL": {
      "type": "string",
      "required": false
    }
  },
  "validations": [],
  "relations": {},
  "acls": [],
  "methods": {}
}`

and on the same page I am getting all linked coming from the database. This has much more to do with the replaceById() method. model-config.json is fine. What I want to do is replace the link fields.
Should I use another method? 

Regards,

**Comment by PhilippeCorreges** [comment]
hi @jonathan-casarrubias 
It appears that the replaceById() method launches http://0.0.0.0:3000/api/Links/58666a91f0d8278a10c038ed/**replace** url. Is this something it should do?
Thanks and of course, Have a Great Year!!

**Comment by jonathan-casarrubias** [comment]
Hi @PhilippeCorreges thanks!!! and have a great year you too!!.

Hey also, it does call for **replace** with asterisk and all that? that seems like a strange url

**Comment by PhilippeCorreges** [comment]
nope the asterisk for for bold in github.
The url is http://0.0.0.0:3000/api/Links/58667c40871eb9e517a4b38e/replace with the same error message : 
`Subscriber.js:227 Uncaught Objectmessage: "Cannot call Link.replaceById(). The replaceById method has not been setup. The PersistedModel has not been correctly attached to a DataSource!"name: "Error"stack: "Error: Cannot call Link.replaceById(). The replaceById method has not been setup. The PersistedModel has not been correctly attached to a DataSource!↵    at throwNotAttached (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback/lib/persisted-model.js:68:11)↵    at Function.replaceById (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback/lib/persisted-model.js:558:5)↵    at SharedMethod.invoke (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/shared-method.js:263:25)↵    at HttpContext.invoke (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/http-context.js:389:12)↵    at phaseInvoke (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/remote-objects.js:654:9)↵    at runHandler (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/lib/phase.js:135:5)↵    at iterate (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/node_modules/async/lib/async.js:146:13)↵    at Object.async.eachSeries (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/node_modules/async/lib/async.js:162:9)↵    at runHandlers (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/lib/phase.js:144:13)↵    at iterate (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/node_modules/async/lib/async.js:146:13)↵    at /Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/node_modules/async/lib/async.js:157:25↵    at /Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/node_modules/async/lib/async.js:154:25↵    at execStack (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/remote-objects.js:493:7)↵    at RemoteObjects.execHooks (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/remote-objects.js:497:10)↵    at phaseBeforeInvoke (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/remote-objects.js:650:10)↵    at runHandler (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/lib/phase.js:135:5)"status: 500__proto__: Object
SafeSubscriber.__tryOrUnsub @ Subscriber.js:227
SafeSubscriber.next @ Subscriber.js:172
Subscriber._next @ Subscriber.js:125
Subscriber.next @ Subscriber.js:89
Subject.next @ Subject.js:55
EventEmitter.emit @ async.js:74
NgZone.triggerError @ ng_zone.js:278
onHandleError @ ng_zone.js:257
ZoneDelegate.handleError @ zone.js:236
Zone.runTask @ zone.js:157
ZoneTask.invoke @ zone.js:335
XMLHttpRequest.send (async)
scheduleTask @ zone.js:1382
ZoneDelegate.scheduleTask @ zone.js:245
Zone.scheduleMacroTask @ zone.js:171
(anonymous) @ zone.js:1405
send @ VM557:3
(anonymous) @ xhr_backend.js:117
Observable.subscribe @ Observable.js:56
Observable._subscribe @ Observable.js:114
MapOperator.call @ map.js:54
Observable.subscribe @ Observable.js:53
Observable._subscribe @ Observable.js:114
CatchOperator.call @ catch.js:30
Observable.subscribe @ Observable.js:53
Observable._subscribe @ Observable.js:114
MapOperator.call @ map.js:54
Observable.subscribe @ Observable.js:53
EditLinkComponent.submit @ edit-link.component.ts:64
View_EditLinkComponent0.handleEvent_12 @ component.ngfactory.js:319
(anonymous) @ view.js:408
(anonymous) @ dom_renderer.js:276
ZoneDelegate.invokeTask @ zone.js:265
onInvokeTask @ ng_zone.js:227
ZoneDelegate.invokeTask @ zone.js:264
Zone.runTask @ zone.js:154
ZoneTask.invoke @ zone.js:335`

**Comment by jonathan-casarrubias** [comment]
Ok, I will add a unit test for this specific all, just to make sure...

Which version of loopback are you using

**Comment by PhilippeCorreges** [comment]
It appears it does a Get ...
I am using: 
**"loopback": "^2.22.0",
    "loopback-boot": "^2.6.5",
    "loopback-component-explorer": "^2.2.0",
    "loopback-component-passport": "2.1.0",
    "loopback-context": "^1.0.0",
    "loopback-datasource-juggler": "2.44.0",**

and     **"@mean-expert/loopback-sdk-builder": "^2.1.0-rc.6",**

**Comment by jonathan-casarrubias** [comment]
Hey @PhilippeCorreges thanks for all of this information, I will be adding a new unit test for this method and see if I can replicate, I believe that currently there is no test for this method, so let me create one and see the results.

Cheers
Jon

**Comment by PhilippeCorreges** [comment]
thanks @jonathan-casarrubias

**Comment by PhilippeCorreges** [comment]
Should i use another method in the meanwhile? I would just like to update a record.

**Comment by jonathan-casarrubias** [comment]
the updateAttributes or upsert methods should be fine for now

**Comment by PhilippeCorreges** [comment]
What do you mean by should be ? Do I have to npm update ?

**Comment by jonathan-casarrubias** [comment]
These methods should be already generated within each model service, just look for these, and I mean these methods should work for the purpose you are looking for.

**Comment by PhilippeCorreges** [comment]
Ok. I ll tell you tomorrow. I think they are not generated as I remember having looked at already.
I have relationships between two models (link is one of those two). If a model has no relationship, the methods are generated.

**Comment by PhilippeCorreges** [comment]
hi @jonathan-casarrubias 
As promised, an update.
I suppressed my sdk folder and generated it again using the lb-ng command. If i go into /shared/sdk/services/custom and edit my Link.ts, I have this:
`export class LinkApi extends BaseLoopBackApi {

  constructor(
    @Inject(Http) protected http: Http,
    @Inject(SocketConnections) protected connections: SocketConnections,
    @Inject(SDKModels) protected models: SDKModels,
    @Inject(LoopBackAuth) protected auth: LoopBackAuth,
    @Inject(JSONSearchParams) protected searchParams: JSONSearchParams,
    @Optional() @Inject(ErrorHandler) protected errorHandler: ErrorHandler
  ) {
    super(http,  connections,  models, auth, searchParams, errorHandler);
  }

  /**
   * Fetches belongsTo relation category.
   *
   * @param any id PersistedModel id
   *
   * @param boolean refresh 
   *
   * @returns object An empty reference that will be
   *   populated with the actual data once the response is returned
   *   from the server.
   *
   * <em>
   * (The remote method definition does not provide any description.
   * This usually means the response is a `Link` object.)
   * </em>
   */
  public getCategory(id: any, refresh: any = {}): Observable<any> {
    let _method: string = "GET";
    let _url: string = LoopBackConfig.getPath() + "/" + LoopBackConfig.getApiVersion() +
    "/Links/:id/category";
    let _routeParams: any = {
      id: id
    };
    let _postBody: any = {};
    let _urlParams: any = {};
    if (refresh) _urlParams.refresh = refresh;
    let result = this.request(_method, _url, _routeParams, _urlParams, _postBody);
    return result;
  }

  /**
   * The name of the model represented by this $resource,
   * i.e. `Link`.
   */
  public getModelName() {
    return "Link";
  }
}`

I have nothing else.
Did I do something wrong ?
I even tried to uninstall the generator with npm uninstall, npm cache clear, npm install --save-dev. Same result.
What I am doing is going to my Loopback project (Not angular2 folder! but the root of my project) and launch: 
`**./node_modules/.bin/lb-sdk server/server.js angular2client/src/app/shared/sdk**`

As the sdk is generated, I imagine the command line is ok.

No idea at all :(

**Comment by jonathan-casarrubias** [comment]
Hi @PhilippeCorreges actually I gave you a wrong path.. I apologize

The methods are not generated within each of the services, instead the PersistedModel methods like updateAttributes are generic and extended from the base service, you can find it within the core directory.

All of your persisted methods should be in there, so again... You should be able to use these methods from your custom services because these extends from the base service.

Cheers
Jon

**Comment by PhilippeCorreges** [comment]
@jonathan-casarrubias 
yep of course IDE code completion helps me on that as it proposes the method.
Now, and this is still my problem, I have this error:
**POST http://0.0.0.0:3000/api/Links/58667c40871eb9e517a4b38e/replace 
500 (Internal Server Error)** 
**Cannot call Link.replaceById(). The replaceById method has not been setup. The PersistedModel has not been correctly attached to a DataSource!"**

Maybe I am doig something bad : 
`this.link.replaceById(this.linkId).subscribe((persistedInstance: Link) => {
        });`

with **private link: LinkApi**

**Comment by PhilippeCorreges** [comment]
ok.
loopback phase was giving the error message so ...
I had a afterRemote() hook. 
There is no more issue with the Persisted Model.

**Comment by jonathan-casarrubias** [comment]
Awesome @PhilippeCorreges I'm glad you found the issue and made it work, these type of problems makes you expert in some matter. 

Actually in LoopBack that is a great tip: When a endpoint or method does not work, make sure a hook is not breaking it.

cheers
Jon

**Comment by PhilippeCorreges** [comment]
FYI,
It does not work in fact ...
and seems not to be linked to remote hook as I suppressed it and restarted the server.
I just did the operation in loopback API explorer app and get that same issue:
{
  "error": {
    "name": "Error",
    "status": 500,
    "message": "Cannot call Link.replaceById(). The replaceById method has not been setup. The PersistedModel has not been correctly attached to a DataSource!",
    "stack": "Error: Cannot call Link.replaceById(). The replaceById method has not been setup. The PersistedModel has not been correctly attached to a DataSource!\n    at throwNotAttached (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback/lib/persisted-model.js:68:11)\n    at Function.replaceById (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback/lib/persisted-model.js:558:5)\n    at SharedMethod.invoke (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/shared-method.js:263:25)\n    at HttpContext.invoke (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/http-context.js:389:12)\n    at phaseInvoke (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/remote-objects.js:654:9)\n    at runHandler (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/lib/phase.js:135:5)\n    at iterate (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/node_modules/async/lib/async.js:146:13)\n    at Object.async.eachSeries (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/node_modules/async/lib/async.js:162:9)\n    at runHandlers (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/lib/phase.js:144:13)\n    at iterate (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/node_modules/async/lib/async.js:146:13)\n    at /Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/node_modules/async/lib/async.js:157:25\n    at /Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/node_modules/async/lib/async.js:154:25\n    at execStack (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/remote-objects.js:493:7)\n    at RemoteObjects.execHooks (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/remote-objects.js:497:10)\n    at phaseBeforeInvoke (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/strong-remoting/lib/remote-objects.js:650:10)\n    at runHandler (/Users/IOSoftwareSAS/WebstormProjects/distinguo/node_modules/loopback-phase/lib/phase.js:135:5)"
  }
}

Do not know about a loopback module not up to date or even another method to use.
What is your advice?

**Comment by jonathan-casarrubias** [comment]
Actually this seems to be a quite strange issue, I never had this one... and I had many.

Maybe is a good time to raise a ticket in their side for this, if the error remains while using the API Explorer then seems to be out of this scope

**Comment by PhilippeCorreges** [comment]
I know.

**Comment by PhilippeCorreges** [comment]
Mmmm I have : 
npm --version strongloop
3.9.2

Does it talk to you ? Are you still in 2.22.0 ?

**Comment by PhilippeCorreges** [comment]
In fact I think the replace Method is not provided by default with loopback 2 isnt'it?

**Comment by jonathan-casarrubias** [comment]
Yes I'm still using version 2 and I do get replaceById method from the API Explorer? didn't you mention that you were able to see the endpoint in the API Explorer?

**Comment by PhilippeCorreges** [comment]
@jonathan-casarrubias 
After deep investigations :) ...
I used to use your starter project. So I developed my stuff based on it and I went to that '**update issue**'.
So, after posting things there and there, a comment put my eye on a point: My Link Model **extends Persisted Model**. The only model I used for "creation" was a model extending User. So a comment somewhere told me that I was trying to use methods that were not by default.
Strange isn't it?
I thought that was not the case and it was not. Thinking for a while ... And I have seen that I forgot to setup the connector to the database!!!!! Mongo connector in my case. (That corresponds to the error message :) )
Do not ask me why I had a replace() method in API explorer! I absolutely **DO NOT KNOW**.
I set the connector up. replace() method disappeared and I am successfully using updateAttributes() one.

pfouahhhh
